core/src/main/java/com/linecorp/armeria/client/endpoint/RampingUpWeightedRoundRobinStrategy.java

```diff
@@ -0,0 +1,433 @@
+                    // when newlyAddedEndpoints is empty.
+                    endpointSelector = new WeightBasedRandomEndpointSelector(builder.build());
+                    return;
+                }
+
+                if (addToNextEntry()) {
+                    // Combine with the Entry and let it handle this newEndpoints when it calls updateWeight().
+                    unhandledNewEndpoints = newEndpoints;
+                    return;
+                }
+            }
+
+            final Deque<EndpointAndStep> newlyAddedEndpoints = updateEndpoints1(newEndpoints);
+            if (newlyAddedEndpoints.isEmpty()) {
+                final ImmutableList.Builder<Endpoint> builder = ImmutableList.builder();
+                addCurrentEndpoints(builder);
+                // Should recreate endpointSelector because the oldEndpoints is changed
+                // when newlyAddedEndpoints is empty.
+                endpointSelector = new WeightBasedRandomEndpointSelector(builder.build());
+                return;
+            }
+
+            final EndpointsInUpdatingEntry endpointsInUpdatingEntry =
+                    new EndpointsInUpdatingEntry(newlyAddedEndpoints, ticker,
+                                                 TimeUnit.MILLISECONDS.toNanos(rampingUpIntervalMillis));
+            endpointsInUpdatingEntries.add(endpointsInUpdatingEntry);
+            updateEndpointWeight();
+        }
+
+        private boolean addToPrevEntry() {
+            final EndpointsInUpdatingEntry lastEndpointsInUpdatingEntry = endpointsInUpdatingEntries.peekLast();
+            return lastEndpointsInUpdatingEntry != null &&
+                   ticker.read() - lastEndpointsInUpdatingEntry.lastUpdatedTime <= updatingTaskWindowNanos;
+        }
+
+        private boolean addToNextEntry() {
+            final EndpointsInUpdatingEntry nextEndpointsInUpdatingEntry = endpointsInUpdatingEntries.peek();
+            return nextEndpointsInUpdatingEntry != null &&
+                   nextEndpointsInUpdatingEntry.nextUpdatingTime - ticker.read() <= updatingTaskWindowNanos;
+        }
+
+        private void addCurrentEndpoints(Builder<Endpoint> builder) {
+            assert oldEndpoints != null;
+            builder.addAll(oldEndpoints);
+            endpointsInUpdatingEntries.forEach(entry -> {
+                entry.endpointAndSteps().forEach(endpointAndStep -> {
+                    builder.add(endpointAndStep.endpoint().withWeight(endpointAndStep.currentWeight()));
+                });
+            });
+        }
+
+        private Deque<EndpointAndStep> updateEndpoints1(List<Endpoint> newEndpoints) {
+            final List<Endpoint> replacedOldEndpoints = new ArrayList<>();
+            assert oldEndpoints != null;
+            for (final Iterator<Endpoint> i = oldEndpoints.iterator(); i.hasNext();) {
+                final Endpoint oldEndpoint = i.next();
+                for (Endpoint newEndpoint : newEndpoints) {
```

Comment:
We also need to remove an `oldEnpoint` which is not in `newEndpoints` so this solution does not guarantee the reduction of time complexity when endpoints are such a big number.
We also have the same problem in [HealthCheckedEndpoint](https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroup.java#L181) (`contatins` method uses a for loop inside.)

To prevent the situation I think sorting the endpoints and comparing from the first is the best solution when N is such a big number. However, this is not good when N is small and also it will make this PR complicated. So let me just add TODO for that. 😉

Reply:
I now understand what you meant. 😄 Let me use the map. Thanks!